openadapt/ml/finetune.py

```diff
@@ -0,0 +1,110 @@
+from openadapt.crud import *
+from openadapt import models, utils
+from copy import deepcopy
+import json
+from loguru import logger
+
+from openadapt.strategies.stateful import get_window_state_diffs
+
+
+def condense_window_state(recording_id: int):
+    """
+    Takes in a recording as input and outputs a list of tuples
+    of action, window events where window events are condensed
+    to contain basic information, as in stateful_test.py
+    """
+    grab_recording = get_recording_by_id(recording_id)
+    total_acx = get_action_events(grab_recording)
+
+    for i in range(len(total_acx)):
+        processed_acx, processed_wd = (
+            sanitize(total_acx[i])[0],
+            sanitize(total_acx[i])[1],
+        )
+        processed_wd.pop("meta")
+        # task_description = recording.task_description
+
+        logger.debug(f"{total_acx[i].timestamp=}")
+        logger.debug(f"{total_acx[i].window_event_timestamp=}")
+        # which timestamp to finetune on?
+        total_acx[i] = (processed_acx, processed_wd)
+    return total_acx
+
+
+def write_to_file(recording_id: int):
+    condensed_recording = condense_window_state(recording_id)
+    # for curr_acx, curr_window in condensed_recording:
+    #   curr_pair= f'action:{curr_acx}, window: {curr_window}'
+    #   pair_json = json.loads(curr_pair)
+    with open(
+        "/Users/owaiszahid/Desktop/ctx/OpenAdapt/openadapt/ml/prompt_completion.json",
+        mode="a",
+    ) as json_file:
+        for idx in range(len(condensed_recording) - 1):
+            curr_acx = condensed_recording[idx][0][0]
+            curr_win = condensed_recording[idx][1]
+
+            next_acx = condensed_recording[idx + 1][0][0]
+            next_win = condensed_recording[idx + 1][1]
+            paired_dict = {
+                "prompt": f"{(curr_acx, curr_win)}",
+                "completion": f"{(next_acx,next_win)}",
+            }
+            # write this to a file
+            # paired_dict_json = json.loads(str(paired_dict))
+            json_file.write(json.dumps(paired_dict))
+            json_file.write("\n")
+
+        json_file.close()
+
+
+def finetune(recording_id: int):
```

Comment:
Usually you run CLI commands for OpenAI model finetuning, but this is blank because finetuning on each model is different.